File: shared/domain/src/family.ts
// The Vogel Vault — shared family/visibility contract.
//
// This is a direct port of MasonsBudget/MasonsBudget/Models/SharedEnums.swift.
// The Swift enum is the source of truth; this file must stay behaviourally
// identical to it. Parity is enforced by fixtures/visibility-cases.json, which
// the TypeScript and Kotlin test suites both consume.
//
// HARD RULE (repo AGENTS.md): Victor and Rachel are ONE shared household and see
// identical data. Canonical untagged adult records default to owner "victor",
// so any strict `owner === activeMember` check empties Rachel's tabs. That bug
// shipped in v0.3. Always route visibility through canSeeDataOwnedBy — never
// strict equality.

export const FAMILY_MEMBERS = ["victor", "rachel", "mason", "maddox"] as const

export type FamilyMember = (typeof FAMILY_MEMBERS)[number]

export function isFamilyMember(value: unknown): value is FamilyMember {
  return typeof value === "string" && (FAMILY_MEMBERS as readonly string[]).includes(value)
}

/** Untagged adult records resolve to "victor", matching the Swift default. */
export const DEFAULT_OWNER: FamilyMember = "victor"

export function coerceOwner(value: unknown): FamilyMember {
  return isFamilyMember(value) ? value : DEFAULT_OWNER
}

const ADULTS: ReadonlySet<FamilyMember> = new Set<FamilyMember>(["victor", "rachel"])

export function isAdult(member: FamilyMember): boolean {
  return ADULTS.has(member)
}

/**
 * Canonical owner for financial ledger records.
 *
 * Victor and Rachel are one adult household whose durable records are stored
 * as Victor-owned. The active profile remains the actor/viewer; it must never
 * be copied into the owner field merely because Rachel initiated a write.
 */
export function ledgerOwner(member: FamilyMember): FamilyMember {
  return isAdult(member) ? DEFAULT_OWNER : member
}

/** Adults get the full budget/spending surface; kids get a Bitcoin-focused one. */
export function showsFullBudget(member: FamilyMember): boolean {
  return isAdult(member)
}

/**
 * Adults can see the household and the kids. Kids see only their own data.
 * Port of `FamilyMember.canSee(dataOwnedBy:)`.
 */
export function canSeeDataOwnedBy(viewer: FamilyMember, owner: FamilyMember): boolean {
  if (viewer === owner) return true
  if (isAdult(viewer)) return true
  return false
}

/**
 * Net-worth totals are household-scoped for adults, but child stacks must never
 * roll into adult totals. Port of `FamilyMember.sharesNetWorth(with:)`.
 *
 * Note this is deliberately narrower than canSeeDataOwnedBy: an adult can *see*
 * Mason's Strike balance on his profile, but it is not part of adult net worth.
 */
export function sharesNetWorthWith(viewer: FamilyMember, owner: FamilyMember): boolean {
  if (viewer === owner) return true
  return isAdult(viewer) && isAdult(owner)
}

/** Kids cannot switch out of their own profile. Port of `allowedSwitchTargets`. */
export function allowedSwitchTargets(member: FamilyMember): FamilyMember[] {
  return isAdult(member) ? [...FAMILY_MEMBERS] : [member]
}

/** Every profile switch is authenticated. Port of `requiresAuthToSwitch`. */
export function requiresAuthToSwitch(_member: FamilyMember): boolean {
  return true
}

export function displayName(member: FamilyMember): string {
  return member.charAt(0).toUpperCase() + member.slice(1)
}

export function profileDescription(member: FamilyMember): string {
  switch (member) {
    case "victor":
    case "rachel":
      return "Full budget, spending & Bitcoin"
    case "mason":
      return "Bitcoin stack & spending"
    case "maddox":
      return "Bitcoin stack & allowance"
  }
}

/**
 * Retained blob routing for the approval-gated whole-file writeback path.
 * Runtime clients read row tables and must not use this as a read plan.
 */
export function transactionsDataFileName(member: FamilyMember): string {
  switch (member) {
    case "victor":
    case "rachel":
      return "transactions"
    case "mason":
      return "mason-transactions"
    case "maddox":
      return "maddox-transactions"
  }
}

/** Retained buy-blob routing for approval-gated whole-file writeback. */
export function btcBuysDataFileName(member: FamilyMember): string {
  return member === "mason" ? "mason-bitcoin-buys" : "bitcoin-buys"
}

/** Mason has dedicated child finance blobs; Maddox currently has none. */
export function hasDedicatedChildFinanceFiles(member: FamilyMember): boolean {
  return member === "mason"
}

// ── Collection helpers ──────────────────────────────────────────────────────
// The clients should filter through these rather than hand-rolling predicates,
// so the v0.3 strict-equality regression cannot reappear per-view.

export interface Owned {
  readonly owner: FamilyMember
}

export function visibleTo<T extends Owned>(viewer: FamilyMember, items: readonly T[]): T[] {
  return items.filter((item) => canSeeDataOwnedBy(viewer, item.owner))
}

export function netWorthScopeFor<T extends Owned>(viewer: FamilyMember, items: readonly T[]): T[] {
  return items.filter((item) => sharesNetWorthWith(viewer, item.owner))
}

File: android/domain/src/main/kotlin/com/sats21m/vogelvault/domain/Family.kt
package com.sats21m.vogelvault.domain

/**
 * The Vogel Vault — shared family/visibility contract, Kotlin mirror.
 *
 * This is a direct port of `MasonsBudget/MasonsBudget/Models/SharedEnums.swift`,
 * which stays the source of truth for all three clients. Parity with Swift and
 * TypeScript is enforced by `shared/domain/fixtures/visibility-cases.json`,
 * which this module's test suite and the TypeScript suite both load.
 *
 * HARD RULE (repo AGENTS.md): Victor and Rachel are ONE shared household and see
 * identical data. Canonical adult records default to owner "victor", so any strict
 * `owner == activeMember` check empties Rachel's tabs. That bug shipped in v0.3.
 * Always route visibility through [canSee] — never strict equality.
 */
enum class FamilyMember(val key: String) {
    VICTOR("victor"),
    RACHEL("rachel"),
    MASON("mason"),
    MADDOX("maddox");

    val isAdult: Boolean
        get() = this == VICTOR || this == RACHEL

    /** Canonical owner for financial records; the active profile remains the actor/viewer. */
    val ledgerOwner: FamilyMember
        get() = if (isAdult) VICTOR else this

    /** Adults get the full budget surface; kids get a Bitcoin-focused one. */
    val showsFullBudget: Boolean
        get() = isAdult

    val displayName: String
        get() = key.replaceFirstChar { it.uppercase() }

    val profileDescription: String
        get() = when (this) {
            VICTOR, RACHEL -> "Full budget, spending & Bitcoin"
            MASON -> "Bitcoin stack & spending"
            MADDOX -> "Bitcoin stack & allowance"
        }

    /** Kids cannot switch out of their own profile. */
    val allowedSwitchTargets: List<FamilyMember>
        get() = if (isAdult) entries.toList() else listOf(this)

    /** Eligible destinations behind the Android profile authentication gate. */
    val gatedSwitchTargets: List<FamilyMember>
        get() = if (isAdult) allowedSwitchTargets else entries.filter { it == this || it.isAdult }

    /** Every profile switch is authenticated in the shared baseline contract. */
    val requiresAuthToSwitch: Boolean
        get() = true

    val transactionsDataFileName: String
        get() = when (this) {
            VICTOR, RACHEL -> "transactions"
            MASON -> "mason-transactions"
            MADDOX -> "maddox-transactions"
        }

    val btcBuysDataFileName: String
        get() = if (this == MASON) "mason-bitcoin-buys" else "bitcoin-buys"

    val hasDedicatedChildFinanceFiles: Boolean
        get() = this == MASON

    /** Adults can see the household and the kids. Kids see only their own data. */
    fun canSee(dataOwnedBy: FamilyMember): Boolean {
        if (this == dataOwnedBy) return true
        if (isAdult) return true
        return false
    }

    /**
     * Net-worth totals are household-scoped for adults, but child stacks must
     * never roll into adult totals.
     *
     * Deliberately narrower than [canSee]: an adult can see Mason's balance on
     * his profile, but it is not part of adult net worth.
     */
    fun sharesNetWorth(with: FamilyMember): Boolean {
        if (this == with) return true
        return isAdult && with.isAdult
    }

    companion object {
        /** Untagged adult records resolve to "victor". Mirrors the Swift default. */
        val DEFAULT_OWNER: FamilyMember = VICTOR

        fun fromKeyOrNull(key: String?): FamilyMember? = entries.firstOrNull { it.key == key }

        /** Unknown or absent owners fall back to the canonical default rather than throwing. */
        fun coerceOwner(key: String?): FamilyMember = fromKeyOrNull(key) ?: DEFAULT_OWNER
    }
}

/** Anything the visibility rules can be applied to. */
interface Owned {
    val owner: FamilyMember
}

/**
 * Filter to what [viewer] may see. Clients should use this rather than
 * hand-rolling a predicate per screen, so the v0.3 strict-equality regression
 * cannot reappear one view at a time.
 */
fun <T : Owned> List<T>.visibleTo(viewer: FamilyMember): List<T> =
    filter { viewer.canSee(it.owner) }

/** Filter to what counts toward [viewer]'s net worth. Narrower than [visibleTo]. */
fun <T : Owned> List<T>.netWorthScopeFor(viewer: FamilyMember): List<T> =
    filter { viewer.sharesNetWorth(it.owner) }
